### ISSUE TITLE
java.lang.NullPointerException in com.databricks.spark.avro.DefaultSource

### ISSUE THREAD

**Issue by liangzhaozeng** [databricks/spark-avro#147]
In Spark 2.0, while reading avro file, got  java.lang.NullPointerException exception in com.databricks.spark.avro.DefaultSource. Here is code snippet that throws exception 

```scala
   val sc = new SparkContext(conf)
   val inStream = "/log/*.avro"
    val sqlContext = new SQLContext(sc)
    val spark = SparkSession.builder()
    println("Loading file " + inStream)
    val df: DataFrame = sqlContext.read.format("com.databricks.spark.avro").option("mergeSchema", "true").load(inStream)
    df.printSchema()
    val selectDF: DataFrame =
    df.select(df(“text"),df(“id"))
    val rdd1 = selectDF.rdd.map(r => (r.getString(0), r.getString(1)))
    println("Loading " + rdd1.count + " records from " + inStream)

 (0 + 17) / 21]16/08/02 22:48:42 WARN scheduler.TaskSetManager: Lost task 18.0 in stage 0.0 (TID 4, [REDACTED]): java.lang.NullPointerException
	at com.databricks.spark.avro.DefaultSource$$anonfun$buildReader$1.apply(DefaultSource.scala:151)
	at com.databricks.spark.avro.DefaultSource$$anonfun$buildReader$1.apply(DefaultSource.scala:143)
	at org.apache.spark.sql.execution.datasources.FileFormat$$anon$1.apply(fileSourceInterfaces.scala:279)
	at org.apache.spark.sql.execution.datasources.FileFormat$$anon$1.apply(fileSourceInterfaces.scala:263)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:116)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:91)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:370)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at org.apache.spark.util.Utils$.getIteratorSize(Utils.scala:1682)
	at org.apache.spark.rdd.RDD$$anonfun$count$1.apply(RDD.scala:1115)
	at org.apache.spark.rdd.RDD$$anonfun$count$1.apply(RDD.scala:1115)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1897)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1897)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:70)
	at org.apache.spark.scheduler.Task.run(Task.scala:85)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:274)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by JoshRosen** [comment]
Do you happen to be using Kryo serialization? If so, I think there's a bug where `spark-avro`'s copy of `SerializableConfiguration` isn't properly registered with Spark's Kryo, causing it to be mis-serialized because its readObject / writeObject methods aren't being invoked.

**Comment by liangzhaozeng** [comment]
Yes, I do. Let me try the fix later on.

**Comment by liangzhaozeng** [comment]
Here are exception after apply this fix.

[Stage 0:>                                                        (0 + 11) / 21]
[Stage 0:>                                                        (0 + 12) / 21]
[Stage 0:>                                                        (0 + 13) / 21]16/08/03 18:25:10 WARN scheduler.TaskSetManager: Lost task 18.0 in stage 0.0 (TID 1, ulnode373.pv05.siri.apple.com): java.io.IOException: com.esotericsoftware.kryo.KryoException: Error during Java deserialization.
	at org.apache.spark.util.Utils$.tryOrIOException(Utils.scala:1260)
	at org.apache.spark.broadcast.TorrentBroadcast.readBroadcastBlock(TorrentBroadcast.scala:174)
	at org.apache.spark.broadcast.TorrentBroadcast._value$lzycompute(TorrentBroadcast.scala:65)
	at org.apache.spark.broadcast.TorrentBroadcast._value(TorrentBroadcast.scala:65)
	at org.apache.spark.broadcast.TorrentBroadcast.getValue(TorrentBroadcast.scala:89)
	at org.apache.spark.broadcast.Broadcast.value(Broadcast.scala:70)
	at com.databricks.spark.avro.DefaultSource$$anonfun$buildReader$1.apply(DefaultSource.scala:146)
	at com.databricks.spark.avro.DefaultSource$$anonfun$buildReader$1.apply(DefaultSource.scala:145)
	at org.apache.spark.sql.execution.datasources.FileFormat$$anon$1.apply(fileSourceInterfaces.scala:279)
	at org.apache.spark.sql.execution.datasources.FileFormat$$anon$1.apply(fileSourceInterfaces.scala:263)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:116)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:91)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:370)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at org.apache.spark.util.Utils$.getIteratorSize(Utils.scala:1682)
	at org.apache.spark.rdd.RDD$$anonfun$count$1.apply(RDD.scala:1115)
	at org.apache.spark.rdd.RDD$$anonfun$count$1.apply(RDD.scala:1115)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1897)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1897)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:70)
	at org.apache.spark.scheduler.Task.run(Task.scala:85)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:274)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: com.esotericsoftware.kryo.KryoException: Error during Java deserialization.
	at com.esotericsoftware.kryo.serializers.JavaSerializer.read(JavaSerializer.java:65)
	at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:790)
	at org.apache.spark.serializer.KryoDeserializationStream.readObject(KryoSerializer.scala:229)
	at org.apache.spark.broadcast.TorrentBroadcast$$anonfun$10.apply(TorrentBroadcast.scala:254)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1287)
	at org.apache.spark.broadcast.TorrentBroadcast$.unBlockifyObject(TorrentBroadcast.scala:255)
	at org.apache.spark.broadcast.TorrentBroadcast$$anonfun$readBroadcastBlock$1.apply(TorrentBroadcast.scala:189)
	at org.apache.spark.util.Utils$.tryOrIOException(Utils.scala:1253)
	... 28 more
Caused by: java.lang.ClassNotFoundException: com.databricks.spark.avro.DefaultSource$SerializableConfiguration
	at java.net.URLClassLoader$1.run(URLClassLoader.java:366)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:278)
	at java.io.ObjectInputStream.resolveClass(ObjectInputStream.java:625)
	at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1612)
	at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1517)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1771)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1350)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:370)
	at com.esotericsoftware.kryo.serializers.JavaSerializer.read(JavaSerializer.java:63)
	... 35 more

**Comment by liangzhaozeng** [comment]
First, it works if I disable the Kryo serialization. 

I use gradle to compile my code. After I get the jar from sbt assembly, I copy the jar file to lib directory and using following to include it.
compile files('lib/spark-avro-assembly-3.0.1-SNAPSHOT.jar')

then in the program,  i do


import com.databricks.spark.avro._

then 

  val df: DataFrame = sqlContext.read.avro(inStream)

    df.printSchema()

    val selectDF: DataFrame =  df.select(df("inputText"),df("interactionId"))

    val rdd1 = selectDF.rdd.map(r => (r.getString(0), r.getString(1)))

    println("Loading " + rdd1.count + " records from " + inStream)

**Comment by JoshRosen** [comment]
Strange. Assuming that you're not just running this in `local` mode, how is the `spark-avro` assembly JAR winding up on the executor classpath? Are you using `spark-submit` with `--jars`?

**Comment by liangzhaozeng** [comment]
I am running in a Yarn cluster and here is my spark-submit


spark-submit \
    --class ${mainclass} \
    --deploy-mode ${deploymode} \
    --master ${SPARK_MASTER} \
    --conf "spark.executor.extraJavaOptions=-XX:+UseG1GC -XX:InitiatingHeapOccupancyPercent=35 -XX:ConcGCThreads=8" \
    --conf spark.yarn.historyServer.address=$SPARK_HISTORY_SERVER \
    --conf spark.eventLog.dir=hdfs://nameservice1/user/spark/applicationHistory \
    --conf spark.eventLog.enabled=true \
    --conf spark.task.maxFailures=64 \
    --conf spark.executor.extraJavaOptions="-XX:PermSize=512m" \
    --driver-java-options -XX:MaxPermSize=512m \
    --driver-memory 8G \
    --executor-cores 5 \
    --executor-memory 8G \
    ${sparkargs} \
    ${mainjar} \
    $*

**Comment by digital-thinking** [comment]
Hi!
I have the same issue, after i ported my spark 1.6.1 application to spark 2.0, while using spark-avro. Also running spark on yarn and using kyro. If you want to know some more details, please feel free to ask.

> 16/08/09 02:38:49 ERROR executor.Executor: Exception in task 40.0 in stage 0.0 (TID 32)
java.lang.NullPointerException
	at com.databricks.spark.avro.DefaultSource$$anonfun$buildReader$1.apply(DefaultSource.scala:151)
	at com.databricks.spark.avro.DefaultSource$$anonfun$buildReader$1.apply(DefaultSource.scala:143)
	at org.apache.spark.sql.execution.datasources.FileFormat$$anon$1.apply(fileSourceInterfaces.scala:279)
	at org.apache.spark.sql.execution.datasources.FileFormat$$anon$1.apply(fileSourceInterfaces.scala:263)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:116)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:91)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:370)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at org.apache.spark.util.collection.ExternalSorter.insertAll(ExternalSorter.scala:192)
	at org.apache.spark.shuffle.sort.SortShuffleWriter.write(SortShuffleWriter.scala:63)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:79)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:47)
	at org.apache.spark.scheduler.Task.run(Task.scala:85)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:274)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)

BTW: Thanks for your awesome work.

**Comment by JoshRosen** [comment]
Hi @digital-thinking, I think that the NPE should have been resolved by #148 but it looks like @liangzhaozeng is hitting a new ClassNotFoundException bug following that fix. We still need to investigate that new error to make sure that this will actually work. If you have any spare testing cycles, it'd be great if you could build a snapshot release from master and try that out to see if you also hit the ClassNotFoundException.

**Comment by digital-thinking** [comment]
Thanks @JoshRosen, i will figure it out.

**Comment by digital-thinking** [comment]
So now i build the snapshot and tested it and ended up with the same ClassNotFoundException. I deactivated Kyro-serialization and it worked. It´s also working locally.

**Comment by koertkuipers** [comment]
i just also ran into the ClassNotFoundException: com.databricks.spark.avro.DefaultSource$SerializableConfiguration
running spark 2.0.0 on yarn with a fat jar that includes spark-avro
spark avro is build by me from master (b078cca6d7b7b758a75eaec6ab1dfc3fafc1c368)
for me it also works if i turn kryo serialization off

**Comment by smallcrew** [comment]
Hi @JoshRosen , just to let you know that the classNotFound issue can be solved using spark configuration, in our case spark.executor.extraClassPath=some.jar
I'm using spark-avro based on revision b078cca

We are using aws emr so we add a bootstrap action to copy our fat jar in our server instances.

**Comment by liangzhaozeng** [comment]
@smallcrew can you tell me what  spark.executor.extraClassPath point to? such like 
spark.executor.extraClassPath=spark-avro-assembly-3.0.1-SNAPSHOT.jar ?

**Comment by smallcrew** [comment]
@liangzhaozeng in our case we just put our fat jar that contains spark-avro classes

**Comment by liangzhaozeng** [comment]
@smallcrew I had tried the extraClassPath but still have same exception before. As you may know, I am running spark 2.0 on Yarn. What's yours? on Yarn?


 @JoshRosen Any plan to fix this bug in short term?

**Comment by fooblahblah** [comment]
I also built the spark-avro-assembly-3.0.1-SNAPSHOT.jar using the referenced commit and attempted to use it with the spark-submit --jars option. The exception I get is:

```java
Exception in thread "main" java.lang.IllegalArgumentException: Class is not registered: com.databricks.spark.avro.DefaultSource$SerializableConfiguration
Note: To register this class use: kryo.register(com.databricks.spark.avro.DefaultSource$SerializableConfiguration.class);
        at com.esotericsoftware.kryo.Kryo.getRegistration(Kryo.java:488)
        at com.esotericsoftware.kryo.util.DefaultClassResolver.writeClass(DefaultClassResolver.java:97)
        at com.esotericsoftware.kryo.Kryo.writeClass(Kryo.java:517)
        at com.esotericsoftware.kryo.Kryo.writeClassAndObject(Kryo.java:622)
        at org.apache.spark.serializer.KryoSerializationStream.writeObject(KryoSerializer.scala:195)
        at org.apache.spark.broadcast.TorrentBroadcast$$anonfun$blockifyObject$2.apply(TorrentBroadcast.scala:236)
        at org.apache.spark.broadcast.TorrentBroadcast$$anonfun$blockifyObject$2.apply(TorrentBroadcast.scala:236)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1287)
        at org.apache.spark.broadcast.TorrentBroadcast$.blockifyObject(TorrentBroadcast.scala:237)
        at org.apache.spark.broadcast.TorrentBroadcast.writeBlocks(TorrentBroadcast.scala:107)
        at org.apache.spark.broadcast.TorrentBroadcast.<init>(TorrentBroadcast.scala:86)
        at org.apache.spark.broadcast.TorrentBroadcastFactory.newBroadcast(TorrentBroadcastFactory.scala:34)
        at org.apache.spark.broadcast.BroadcastManager.newBroadcast(BroadcastManager.scala:56)
        at org.apache.spark.SparkContext.broadcast(SparkContext.scala:1370)
        at com.databricks.spark.avro.DefaultSource.buildReader(DefaultSource.scala:144)
        at org.apache.spark.sql.execution.datasources.FileFormat$class.buildReaderWithPartitionValues(fileSourceInterfaces.scala:260)
        at com.databricks.spark.avro.DefaultSource.buildReaderWithPartitionValues(DefaultSource.scala:46)
        at org.apache.spark.sql.execution.datasources.FileSourceStrategy$.apply(FileSourceStrategy.scala:112)
        at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$1.apply(QueryPlanner.scala:60)
```

**Comment by fooblahblah** [comment]
I was actually hitting a slightly different error in my last comment. That was due to having `sparkConf.set("spark.kryo.registrationRequired", "true")`. I tried a couple work arounds, but basically had to edit the spark-avro source and strip the `private[avro]` bit off the `DefaultSource` object and then I was able to register `com.databricks.spark.avro.DefaultSource$SerializableConfiguration`. Now I'm getting the NPE mentioned by @liangzhaozeng:

```
Caused by: java.lang.NullPointerException
        at com.databricks.spark.avro.DefaultSource$$anonfun$buildReader$1.apply(DefaultSource.scala:151)
        at com.databricks.spark.avro.DefaultSource$$anonfun$buildReader$1.apply(DefaultSource.scala:143)
        at org.apache.spark.sql.execution.datasources.FileFormat$$anon$1.apply(fileSourceInterfaces.scala:279)
        at org.apache.spark.sql.execution.datasources.FileFormat$$anon$1.apply(fileSourceInterfaces.scala:263)
        at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:116)
        at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:91)
        at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
        at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
        at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:370)
        at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
        at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:125)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:79)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:47)
        at org.apache.spark.scheduler.Task.run(Task.scala:85)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:274)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

**Comment by smallcrew** [comment]
@liangzhaozeng using yarn on emr version 5.0.0
and 
spark.kryo.registrator=com.some.customRegistrator
spark.serializer=org.apache.spark.serializer.KryoSerializer

**Comment by fooblahblah** [comment]
I had a bunch of false leads yesterday. I was able to make this thing work using the patch referenced above https://github.com/databricks/spark-avro/issues/170. No more changes were needed and I applied this directly to master as of commit b078cca6d7b7b758a75eaec6ab1dfc3fafc1c368. 

Thanks @koertkuipers that was a very helpful fix! 

I'm still adding the patched spark-avro assembly jar to my spark assembly jar as that seems the easiest as far as distributing the change.

I'm testing some other jobs, but thus far things seem to be working.
